### PR TITLE
Refactor/sale

### PR DESCRIPTION
**Pull request by Snowflake8**
Refactor Sale to include a User and a Book, as well as an Image (which is nullable and can be changed easily since for now is only null)
It also includes a merge from PR #108 which explains the big number of fails
## Pull reviewers stats
Stats for the last 180 days:
|                                                                                                                                                               | User            | Total reviews | Median time to review                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   | Total comments |
| ------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | -------------- |
| <a href=https://github.com/Huntsekkern><img src="https://avatars.githubusercontent.com/u/44240415?v=4" width="20"></a>                                        | Huntsekkern     | 12            | [50m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44240415~n~'Huntsekkern)~p~180~r~(~(d~'qpumtn~t~'1745)~(d~'qpt0zy~t~'2wmx)~(d~'qpt88n~t~'33vm)~(d~'qptd1d~t~'29r)~(d~'qptd3y~t~'2cc)~(d~'qpt4tp~t~'2b1)~(d~'qpuo7b~t~'mk)~(d~'qq6uip~t~'3j5)~(d~'qq8faa~t~'qv)~(d~'qq8fiz~t~'bv)~(d~'qqjij5~t~'i6m)~(d~'qrihqg~t~'1wf3)~(d~'qrm9wa~t~'17x)~(d~'qrma4k~t~'1g7)~(d~'qrmajv~t~'1vi)~(d~'qqkomy~t~'z8i)~(d~'qqkozj~t~'zl3)~(d~'qqkp4h~t~'zq1)~(d~'qqvo8r~t~'14d)~(d~'qrmb5j~t~'1cc)~(d~'qrni3f~t~'15))))                                                           | **31**         |
| <a href=https://github.com/Snowflake8><img src="https://avatars.githubusercontent.com/u/44499990?v=4" width="20"></a>                                         | Snowflake8      | **14**        | [3h 8m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499990~n~'Snowflake8)~p~180~r~(~(d~'qpt3s9~t~'2zf8)~(d~'qpx29n~t~'1vgi)~(d~'qq0q7g~t~'5jeb)~(d~'qq6k9y~t~'8hk)~(d~'qq6kh9~t~'8ov)~(d~'qq6r5v~t~'6b)~(d~'qqdm38~t~'1fge)~(d~'qq6ncv~t~'8609)~(d~'qq8el9~t~'y2h)~(d~'qqjh1g~t~'gox)~(d~'qqjh87~t~'7lg)~(d~'qqjlvf~t~'b)~(d~'qqsnpq~t~'qn)~(d~'qqxzb3~t~'ikt)~(d~'qrm4bb~t~'1ljt)~(d~'qrkbo5~t~'88t)~(d~'qrm4eh~t~'1rq)~(d~'qrnhv0~t~'7)~(d~'qro2s6~t~'666))))                                                                                                | 20             |
| <a href=https://github.com/so7fie><img src="https://avatars.githubusercontent.com/u/10992808?v=4" width="20"></a>                                             | so7fie          | 12            | [20h 54m](https://app.flowwer.dev/charts/review-time/~(u~(i~'10992808~n~'so7fie)~p~180~r~(~(d~'qpuy0b~t~'1v)~(d~'qq0p5z~t~'5hby)~(d~'qq0t0v~t~'5m7q)~(d~'qq0t2c~t~'5m97)~(d~'qqew7j~t~'2pkp)~(d~'qqf92s~t~'32fy)~(d~'qqjbho~t~'1ux)~(d~'qqy5l4~t~'5im1)~(d~'qq7qp0~t~'hf)~(d~'qrm6v4~t~'6ai)~(d~'qqkh75~t~'dr0)~(d~'qqrud7~t~'7qx2)~(d~'qr8kh3~t~'1kfr)~(d~'qrm9jc~t~'1npk)~(d~'qrmbez~t~'1ls)~(d~'qrnshx~t~'193))))                                                                                                                                                    | 14             |
| <a href=https://github.com/RaulinN><img src="https://avatars.githubusercontent.com/u/44499915?u=c029696ff2de0902f905fbf35188359e94e08d88&v=4" width="20"></a> | RaulinN         | 12            | [2h 4m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499915~n~'RaulinN)~p~180~r~(~(d~'qptcqf~t~'1yt)~(d~'qpq5rb~t~'5xjz)~(d~'qq73ig~t~'jy6)~(d~'qq7vmj~t~'1t5)~(d~'qpuwak~t~'8pt)~(d~'qq73am~t~'28)~(d~'qqhz4w~t~'87bj)~(d~'qq73x3~t~'8mkh)~(d~'qq821x~t~'lj5)~(d~'qqr6wy~t~'aju)~(d~'qreyev~t~'f)~(d~'qq7qbn~t~'1bz)~(d~'qqj0lo~t~'u5)~(d~'qqjmtu~t~'n2b)~(d~'qrm5d5~t~'2qe)~(d~'qrni37~t~'x))))                                                                                                                                                               | 12             |
| <a href=https://github.com/DGoedtkindtEPFL><img src="https://avatars.githubusercontent.com/u/61052800?v=4" width="20"></a>                                    | DGoedtkindtEPFL | **14**        | [19h 46m](https://app.flowwer.dev/charts/review-time/~(u~(i~'61052800~n~'DGoedtkindtEPFL)~p~180~r~(~(d~'qpuy16~t~'2q)~(d~'qpuoy1~t~'co)~(d~'qpup0b~t~'19at)~(d~'qplma9~t~'1e2x)~(d~'qpuugt~t~'am9h)~(d~'qq6s0w~t~'8gm)~(d~'qq6rfm~t~'fn8)~(d~'qqjjxz~t~'9s4m)~(d~'qq9yj2~t~'1i5c)~(d~'qqjmgc~t~'dy)~(d~'qqrtyv~t~'xlr)~(d~'qqxoq1~t~'21ln)~(d~'qqioj1~t~'2x8a)~(d~'qqionj~t~'2xcs)~(d~'qqioou~t~'2xe3)~(d~'qqipcq~t~'2y1z)~(d~'qqru2n~t~'7qmi)~(d~'qrm4pz~t~'1iw7)~(d~'qrm4za~t~'1j5i)~(d~'qrm5eq~t~'1jky)~(d~'qrmfv4~t~'1u1c)~(d~'qrmhyq~t~'1cp)~(d~'qrnxgy~t~'1wd)))) | 7              |
| <a href=https://github.com/Laniakea1999><img src="https://avatars.githubusercontent.com/u/43450242?v=4" width="20"></a>                                       | Laniakea1999    | 4             | [**20m**](https://app.flowwer.dev/charts/review-time/~(u~(i~'43450242~n~'Laniakea1999)~p~180~r~(~(d~'qpupsq~t~'d4)~(d~'qq7oa5~t~'1n)~(d~'qqeyk1~t~'2rx7)~(d~'qrmbac~t~'1h5))))                                                                                                                                                                                                                                                                                                                                                                                          | 1              |

### CODE REVIEW

File: app/src/main/java/com/github/polybooks/core/database/implementation/DummyBookQuery.kt [hunk 1]
```diff
@@ -25,10 +25,11 @@ class DummyBookQuery(private val books : List<Book> = default_books) : BookQuery
        return DummyBookQuery()
     }
 
-    override fun searchByISBN13(isbn13: String) : BookQuery {
-       return DummyBookQuery()
```
Comment (so7fie) [inline app/src/main/java/com/github/polybooks/core/database/implementation/DummyBookQuery.kt:29]: Comme dans la pr de Dario, garder une fonction pour pouvoir directement faire des query avec juste une string

File: app/src/main/java/com/github/polybooks/core/database/implementation/DummySalesQuery.kt [hunk 2]
```diff
@@ -90,7 +88,7 @@ class DummySalesQuery(private val sale: List<Sale> = default_sale) : SaleQuery{
     override fun searchByISBN13(isbn13: String): SaleQuery {
 //        TODO("Not yet implemented")
         Log.d(TAG, "searchByISBN13 not implemented correctly")
-        return DummySalesQuery(sale.filter { sale -> sale.title == isbn13 })
+        return DummySalesQuery(sale.filter { sale -> sale.book.isbn13 == isbn13 })
 
     }
```
Comment (so7fie) [inline app/src/main/java/com/github/polybooks/core/database/implementation/DummySalesQuery.kt:93]: Tu devrais pas ajouter aussi une fonction qui prend un set comme paramètre et enlever "13" dans le nom de la fonction pour plus de consistency?

File: app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt [hunk 3]
```diff
@@ -187,29 +190,69 @@ class SaleDatabase : SaleDatabase {
...
                 SaleFields.PUBLICATION_DATE.fieldName to sale.date,
-                SaleFields.STATE.fieldName to sale.state
+                SaleFields.STATE.fieldName to sale.state,
+                SaleFields.IMAGE.fieldName to null //TODO change this, image goes elsewhere
         )
     }
 
     override fun addSale(sale: Sale) {
+        if(sale.seller == LocalUser)
+            throw LocalUserException("Cannot add sale as LocalUser")
```
Comment (so7fie) [inline app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt:255]: nice :)

File: app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt [hunk 4]
```diff
@@ -220,13 +263,23 @@ class SaleDatabase : SaleDatabase {
...
+            Log.d("1===================================","${task.result.documents}" )
+            //println("1=========================${task.result.documents}")
+            val result = task.result.documents.filter { document ->
+                Log.d("0===================================","$document" )
                 val s = snapshotToSale(document)
-                s.condition == sale.condition && s.seller == sale.seller && s.date == sale.date
+                Log.d("2===================================","LEA IS HERE" )
+                Log.d("2===================================","$s" )
+                //println("2=========================$s")
+                s.condition == sale.condition /*&& s.seller == sale.seller */&& s.date == sale.date
```
Comment (so7fie) [inline app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt:277]: Pas sûre de savoir ce que c'est "LEA IS HERE", mais bon tu dois savoir ce que tu fais

File: app/src/main/java/com/github/polybooks/utils/BookUtil.kt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+package com.github.polybooks.utils
+
+import com.github.polybooks.core.Book
+
+fun anonymousBook(title: String): Book =
+        Book("", null, title, null, null, null, null, null)
```
Comment (so7fie) [inline app/src/main/java/com/github/polybooks/utils/BookUtil.kt:6]: good idea!

**Review by so7fie — APPROVED** [review]
Just need to adapt functions in DummySalesQuery for consistency

File: app/src/androidTest/java/com/github/polybooks/database/DummySalesQueryTest.kt [hunk 6]
```diff
@@ -23,15 +25,15 @@ class DummySalesQueryTest {
...
+            Sale(anonymousBook("Book5"), LocalUser, 21.00f, BookCondition.WORN, com.google.firebase.Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!), SaleState.CONCLUDED, null),
+            Sale(anonymousBook("Book6"), LocalUser, 29.00f, BookCondition.GOOD, com.google.firebase.Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!), SaleState.ACTIVE, null),
+            Sale(anonymousBook("Book7"), LocalUser, 23.00f, BookCondition.GOOD, com.google.firebase.Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!), SaleState.ACTIVE, null),
+            Sale(anonymousBook("Book8"), LocalUser, 23.66f, BookCondition.NEW, com.google.firebase.Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!), SaleState.ACTIVE, null),
+            Sale(anonymousBook("Book9"), LocalUser, 25.00f, BookCondition.GOOD, com.google.firebase.Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!), SaleState.RETRACTED, null),
```
[the remark this answers is not in the document]
Comment (Snowflake8) [reply app/src/androidTest/java/com/github/polybooks/database/DummySalesQueryTest.kt:36]: Parce que la seule raison d'avoir des Timestamp java est qu'ils sont serializable, ce qui ne va plus etre necessaire

File: app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt (hunk 3 above)
Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt:255]: Thx

File: app/src/main/java/com/github/polybooks/utils/BookUtil.kt (hunk 5 above)
Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/utils/BookUtil.kt:6]: Merci

File: app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt (hunk 4 above)
Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt:277]: Hahaha les traces du debugging

File: app/src/main/java/com/github/polybooks/core/database/implementation/DummyBookQuery.kt (hunk 1 above)
Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/core/database/implementation/DummyBookQuery.kt:29]: Ceci vient de la PR de dario donc je le laisse changer ça

File: app/src/main/java/com/github/polybooks/core/database/implementation/DummySalesQuery.kt (hunk 2 above)
Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/core/database/implementation/DummySalesQuery.kt:93]: Je suis d'accord avec la deuxième partie mais moins sure avec la première. Est-ce que on va avoir besoin de chercher des sales avec plusieurs isbn? Je vais y reflechir

Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/core/database/implementation/DummySalesQuery.kt:93]: Alors apres reflexion, j'ai renommé les instances de isbn13 en isbn, mais quant à la fonction, je pense pas que ca rentre dans le domaine de cette PR. Le but c'est de modifier Sale, et changet ceci impliquerait de beaucoup trop toucher à SaleDatabase et toutes ses implementations. Peut-etre mentionne ça dans la PR #108

File: app/src/androidTest/java/com/github/polybooks/database/DummySalesQueryTest.kt [hunk 7]
```diff
@@ -67,19 +66,19 @@ class DummySalesQueryTest {
 
 
         val q1Result = listOf(
-                Sale("Book5", 6, 21.00f, BookCondition.WORN, Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!.time), SaleState.CONCLUDED),
-                Sale("Book9", 9, 25.00f, BookCondition.GOOD, Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!.time), SaleState.RETRACTED),
+                Sale(anonymousBook("Book5"), LocalUser, 21.00f, BookCondition.WORN, Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!), SaleState.CONCLUDED, null),
+                Sale(anonymousBook("Book9"), LocalUser, 25.00f, BookCondition.GOOD, Timestamp(com.github.polybooks.core.database.implementation.format.parse("2016-05-05")!!), SaleState.RETRACTED, null),
```
[the remark this answers is not in the document]
Comment (Snowflake8) [reply app/src/androidTest/java/com/github/polybooks/database/DummySalesQueryTest.kt:67]: C'est le même type partout

File: app/src/main/java/com/github/polybooks/core/database/implementation/DummySalesQuery.kt (hunk 2 above)
Comment (so7fie) [reply app/src/main/java/com/github/polybooks/core/database/implementation/DummySalesQuery.kt]: ça marche, j'avais pas réalisé que c'était pas directement lié

File: app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt [hunk 8]
```diff
@@ -206,31 +201,21 @@ class SaleDatabase : SaleDatabase {
...
         return Sale(
-                a1,//snapshotToBook(snapshot.get(SaleFields.BOOK.fieldName)!! as HashMap<String, Any>),
-                a2,//snapshotToUser(snapshot.get(SaleFields.SELLER.fieldName)!! as HashMap<String, Any>),
-                a3,//snapshot.getLong(SaleFields.PRICE.fieldName)!!.toFloat(),
+                book,
+                seller,
+                price,
                 BookCondition.valueOf(snapshot.getString(SaleFields.CONDITION.fieldName)!!),
                 Timestamp(snapshot.getTimestamp(SaleFields.PUBLICATION_DATE.fieldName)!!.toDate()),
                 SaleState.valueOf(snapshot.getString(SaleFields.STATE.fieldName)!!),
```
Comment (RaulinN) [inline app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt]: Is there a reason why you left half the fields as "non-variable" (and half the fields pre-computed)?

File: app/src/androidTest/java/com/github/polybooks/database/SaleDatabaseTest.kt [hunk 9]
```diff
@@ -293,29 +295,58 @@ class SaleDatabaseTest {
         assertEquals(0,db.querySales().searchByTitle(saleTest.book.title).getCount().get())
     }
 
-
+    @Ignore
     @Test
```
Comment (RaulinN) [inline app/src/androidTest/java/com/github/polybooks/database/SaleDatabaseTest.kt]: Are you sure you fixed the tests as explained by the commit title? The tests you modified (3 of them) seem to be `@Ignore`d

File: app/src/main/java/com/github/polybooks/FilteringSalesActivity.kt [hunk 10]
```diff
@@ -106,7 +106,7 @@ class FilteringSalesActivity : AppCompatActivity() {
             //---
             //DEBUG query.getAll().thenAccept { list -> Log.d(TAG,list.toString())}
             val intent : Intent = Intent(this, ListSalesActivity::class.java)
-            intent.putExtra(ListSalesActivity.EXTRA_SALE_QUERY, query)
+            intent//.putExtra(ListSalesActivity.EXTRA_SALE_QUERY, query)
```
[the remark this answers is not in the document]
Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/FilteringSalesActivity.kt:109]: C'est par rapport a ce que sophie fait

File: app/src/androidTest/java/com/github/polybooks/database/SaleDatabaseTest.kt (hunk 9 above)
Comment (Snowflake8) [reply app/src/androidTest/java/com/github/polybooks/database/SaleDatabaseTest.kt]: Yes these three tests can be used to insert and delete sales from the firebase database. They always work (no asserts in them)

**Review by RaulinN — APPROVED** [review]
This time LGTM :)

**Pull request merged**